lib/jobs/diag-update-firmware.js

```diff
@@ -0,0 +1,198 @@
+        };
+        this.diagTool = {};
+    }
+    util.inherits(DiagUpdateFirmware, BaseJob);
+
+    /**
+     * @memberOf DiagUpdateFirmware
+     */
+    DiagUpdateFirmware.prototype._run = function() {
+        var self = this;
+
+        return Promise.try(function(){
+            var settings = {host: self.getNodeIp()};
+            self.diagTool = new DiagTool(settings, self.nodeId);
+            return self.diagTool.retrySyncDiscovery(5000, 6);
+        })
+        .then(function(){
+            return self.diagTool.getAllDevices();
+        })
+        .then(function(){
+            return self.diagTool.uploadImageFile(self.localImagePath);
+        })
+        .then(function(){
+            var method = 'update' + _.capitalize(self.firmwareName);
+            return self[method](self.imageName, self.imageMode, self.diagImagePath);
+        })
+        .then(function(){
+            self._done();
+        })
+        .catch(function(err){
+            self._done(err);
+        });
+    };
+
+    /**
+    * Get node IP from context
+    * @return {String} IP address string
+    */
+    DiagUpdateFirmware.prototype.getNodeIp = function(){
+        var ip = this.context.nodeIp;
+        assert.isIP(ip, 4);
+        return ip;
+    };
+
+    /**
+    * Parse image mode for bios firmware update
+    * @param {String/Integer} mode: user selected image mode, mapping as below:
+    *   '0'/0/'fullbios' - full BIOS,
+    *   '1'/1/'bios' - BIOS,
+    *   '2'/2/'uefi' - UEFI,
+    *   '3'/3/'serdes' - serdes,
+    *   '4'/4/'post' - POST,
+    *   '5'/5/'me' - ME
+    * @return {String} mode string, a value of '0' - '5'
+    */
+    DiagUpdateFirmware.prototype.parseBiosMode = function(mode){
+        var self = this;
+        var _mode = parseInt(mode);
+        if (_.isNaN(_mode)){
+            mode = self.biosModeMapping[mode];
+        } else {
+            mode = _(mode).toString();
+        }
+        return mode;
+    };
+
+    /**
+    * Parse image mode for bmc firmware update
+    * @param {String} mode: user selected image mode, mapping as below:
+    *   '0x142'/'ssp' - SSP,
+    *   '0x140'/'bmcapp' - BMC Main APP,
+    *   '0x144'/'bootblock' - Bootblock,
+    *   '0x145'/'adaptivecooling' - Adaptive Cooling,
+    *   '0x5f'/'fullbmc' - BMC full image
+    * @return {String} mode string, a hex number.
+    */
+    DiagUpdateFirmware.prototype.parseBmcMode = function(mode){
+        var self = this;
+        if (!_.startsWith(mode, '0x')){
+            mode = self.bmcModeMapping[mode];
+        }
+        return mode;
+    };
+
+    /**
+    * Update bmc firmware
+    * @param {String} imageName: image name
+    * @param {String} mode: image mode
+    * @param {String} diagImagePath: image file path in diag system
+    * @return {Promise}
+    */
+    DiagUpdateFirmware.prototype.updateBmc = function(imageName, imageMode, diagImagePath){
+        var self = this;
+        imageMode = self.parseBmcMode(imageMode);
```

Comment:
imageMode variable name is the same before and after parse, it's dangerous or difficult to debug